editor/editor_node.cpp

```diff
@@ -4784,6 +4804,31 @@ void EditorNode::_save_docks_to_config(Ref<ConfigFile> p_layout, const String &p
 	for (int i = 0; i < hsplits.size(); i++) {
 		p_layout->set_value(p_section, "dock_hsplit_" + itos(i + 1), hsplits[i]->get_split_offset());
 	}
+
+	// FileSystemDock.
+	p_layout->set_value(p_section, "dock_filesystem_split", FileSystemDock::get_singleton()->get_split_offset());
+	p_layout->set_value(p_section, "dock_filesystem_display_mode", FileSystemDock::get_singleton()->get_display_mode());
+	p_layout->set_value(p_section, "dock_filesystem_file_sort", FileSystemDock::get_singleton()->get_file_sort());
+	p_layout->set_value(p_section, "dock_filesystem_file_list_display_mode", FileSystemDock::get_singleton()->get_file_list_display_mode());
+	PackedStringArray selected_files = FileSystemDock::get_singleton()->get_selected_paths();
+	p_layout->set_value(p_section, "dock_filesystem_selected_paths", selected_files);
+
+	// BFS to save the collapsed state of the tree.
+	TreeItem *root_ti = FileSystemDock::get_singleton()->get_tree_control()->get_root();
+	List<TreeItem *> queue;
+	Vector<String> collapsed_paths;
+	queue.push_back(root_ti);
+	while (queue.size()) {
+		TreeItem *ti = queue.back()->get();
+		queue.pop_back();
+		if (!ti->is_collapsed()) {
```

Comment:
This condition is not enough, because it will store also files. You need to check if it's a directory first.

Reply:
You're right. While it basically works since the collapsed property is ignored for leaves, just saving this information for inner nodes of the tree is more efficient (especially when several hundred files are involved). Instead of checking directlity whether it's a directory, I went with this general approach which might be a bit more future-proof and cleaner:
```suggestion
		if (!ti->is_collapsed() && ti->get_child_count() > 0) {
```